### PR TITLE
[CINFRA-135] Update Participant Flags

### PR DESCRIPTION
**Pull request by maierlars**
### Scope & Purpose

Allow updating participant flags without changing the term. The new configuration is used immediately. However, we write an empty entry and once this entry is committed, we accept the new configuration.

### CODE REVIEW

**Review by markuspf — CHANGES_REQUESTED** [review]
I consider this the first round of reviews where I mostly commented on details that I noticed "obviously".

This PR is hard to review because it consists of a lot of legwork to get the actual task done; this is not necessarily a criticism of the PR or our process itself, but we might want to think about how to make tasks like this one more palatable to implement or break up.

File: arangod/Replication2/ReplicatedLog/LogCommon.h [hunk 1]
```diff
@@ -328,6 +331,37 @@ struct LogConfig {
 [[nodiscard]] auto operator==(LogConfig const& left, LogConfig const& right) noexcept -> bool;
 [[nodiscard]] auto operator!=(LogConfig const& left, LogConfig const& right) noexcept -> bool;
 
+struct ParticipantFlags {
```
[the remark this answers is not in the document]
Comment (maierlars) [reply arangod/Replication2/ReplicatedLog/LogCommon.h:334]: We agreed on the following solution: move the `failed` flag to the `ParticipantStateTuple` and then use the `ParticipantFlags` from `LogCommon.h`. We then only modify the `is*` function to read the correct value.

Then we have to adjust tests.

File: arangod/Replication2/ReplicatedLog/LogLeader.h [hunk 2]
```diff
@@ -143,6 +151,17 @@ class LogLeader : public std::enable_shared_from_this<LogLeader>, public ILogPar
   // entry within its term has been committed.
   [[nodiscard]] auto isLeadershipEstablished() const noexcept -> bool;
 
+  // This function returns the current commit index. Do NOT pull this function,
```
[the remark this answers is not in the document]
Comment (maierlars) [reply arangod/Replication2/ReplicatedLog/LogLeader.h:154]: Typo. `Do NOT poll this function`.

File: arangod/Replication2/ReplicatedLog/LogStatus.h [hunk 3]
```diff
@@ -56,6 +56,8 @@ struct LeaderStatus {
   // now() - insertTP of last uncommitted entry
   std::chrono::duration<double, std::milli> commitLagMS;
   CommitFailReason lastCommitStatus;
+  ParticipantsConfig acceptedParticipantConfig;
+  ParticipantsConfig committedParticipantConfig;
```
[the remark this answers is not in the document]
Comment (maierlars) [reply arangod/Replication2/ReplicatedLog/LogStatus.h:61]: Not yet documented. `accepted` - the follower has accepted to use the configuration from now on to commit log entries. `committed` - at least one log entry was committed using the accepted log configuration.

File: arangod/Replication2/ReplicatedLog/LogStatus.h [hunk 4]
```diff
@@ -96,6 +98,8 @@ struct LogStatus {
   [[nodiscard]] auto getCurrentTerm() const noexcept -> std::optional<LogTerm>;
   [[nodiscard]] auto getLocalStatistics() const noexcept -> std::optional<LogStatistics>;
 
+  [[nodiscard]] auto asLeaderStatus() const noexcept -> LeaderStatus const*;
```
[the remark this answers is not in the document]
Comment (maierlars) [reply arangod/Replication2/ReplicatedLog/LogStatus.h:101]: `std::optional<LeaderStatus const&>` is not available and a pointer is perfect for a nullable reference.

File: arangod/Replication2/ReplicatedLog/LogCommon.cpp [hunk 5]
```diff
@@ -276,24 +276,29 @@ auto replication2::operator<<(std::ostream& os, TermIndexPair pair) -> std::ostr
...
+  builder.add(VPackStringRef(StaticStrings::SoftWriteConcern), VPackValue(softWriteConcern));
   builder.add(StaticStrings::ReplicationFactor, VPackValue(replicationFactor));
 }
 
 auto replication2::operator==(LogConfig const& left, LogConfig const& right) noexcept -> bool {
   // TODO How can we make sure that we never forget a field here?
   return left.waitForSync == right.waitForSync && left.writeConcern == right.writeConcern &&
+         left.softWriteConcern == right.softWriteConcern &&
          left.replicationFactor == right.replicationFactor;
 }
```
[the remark this answers is not in the document]
Comment (maierlars) [reply arangod/Replication2/ReplicatedLog/LogCommon.cpp:304]: yes.

File: arangod/Replication2/ReplicatedLog/LogCommon.h [hunk 6]
```diff
@@ -29,6 +29,7 @@
 #include <string>
 #include <string_view>
 #include <variant>
+#include <unordered_map>
```
[the remark this answers is not in the document]
Comment (maierlars) [reply arangod/Replication2/ReplicatedLog/LogCommon.h:32]: yes

File: arangod/Cluster/Maintenance.cpp [hunk 7]
```diff
@@ -1197,131 +1210,170 @@ static VPackBuilder assembleLocalDatabaseInfo(DatabaseFeature& df, std::string c
   }
 }
 
-static void reportCurrentLocalReplicatedLog(VPackBuilder& report,
+static auto reportCurrentReplicatedLogLocal(replication2::replicated_log::LogStatus const& status,
```
[the remark this answers is not in the document]
Comment (maierlars) [reply arangod/Cluster/Maintenance.cpp:1213]: ähm ... consistency? 🤷🏽‍♂️ That's how you write maintenance code.

File: arangod/Replication2/ReplicatedLog/LogStatus.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (maierlars) [reply arangod/Replication2/ReplicatedLog/LogStatus.h:61]: renamed accepted to `active`.

**Review by markuspf — APPROVED** [review]
this now looks good to me; it bugs me a bit that there seems to be so much unrelated boilerplate that i needed to wade through; let's discuss this tomorrow.

**Pull request merged**